azure-client-runtime/src/main/java/com/microsoft/azure/v2/AzureAsyncOperationPollStrategy.java

```diff
@@ -0,0 +1,160 @@
+ * running operation.
+ */
+public final class AzureAsyncOperationPollStrategy extends PollStrategy {
+    private final String fullyQualifiedMethodName;
+    private final String operationResourceUrl;
+    private final String originalResourceUrl;
+    private final SerializerAdapter<?> serializer;
+    private Boolean pollingSucceeded;
+    private boolean gotResourceResponse;
+
+    /**
+     * The name of the header that indicates that a long running operation will use the
+     * Azure-AsyncOperation strategy.
+     */
+    public static final String HEADER_NAME = "Azure-AsyncOperation";
+
+    /**
+     * The provisioning state of the operation resource if the operation is still in progress.
+     */
+    public static final String IN_PROGRESS = "InProgress";
+
+    /**
+     * The provisioning state of the operation resource if the operation is successful.
+     */
+    public static final String SUCCEEDED = "Succeeded";
+
+    /**
+     * Create a new AzureAsyncOperationPollStrategy object that will poll the provided operation
+     * resource URL.
+     * @param fullyQualifiedMethodName The fully qualified name of the method that initiated the
+     *                                 long running operation.
+     * @param operationResourceUrl The URL of the operation resource this pollStrategy will poll.
+     * @param serializer The serializer that will deserialize the operation resource and the
+     *                   final operation result.
+     */
+    private AzureAsyncOperationPollStrategy(String fullyQualifiedMethodName, String operationResourceUrl, String originalResourceUrl, SerializerAdapter<?> serializer) {
+        this.fullyQualifiedMethodName = fullyQualifiedMethodName;
+        this.operationResourceUrl = operationResourceUrl;
+        this.originalResourceUrl = originalResourceUrl;
+        this.serializer = serializer;
+    }
+
+    @Override
+    public HttpRequest createPollRequest() {
+        String pollUrl = null;
+        if (pollingSucceeded == null) {
+            pollUrl = operationResourceUrl;
+        }
+        else if (pollingSucceeded) {
+            pollUrl = originalResourceUrl;
+        }
+        return new HttpRequest(fullyQualifiedMethodName, "GET", pollUrl);
+    }
+
+    @Override
+    public void updateFrom(HttpResponse httpPollResponse) throws IOException {
+        updateDelayInMillisecondsFrom(httpPollResponse);
+
+        if (pollingSucceeded == null) {
+            final String bodyString = httpPollResponse.bodyAsString();
+            updateFromResponseBodyString(bodyString);
+        }
+        else if (pollingSucceeded) {
+            gotResourceResponse = true;
+        }
+    }
+
+    @Override
+    public Single<HttpResponse> updateFromAsync(final HttpResponse httpPollResponse) {
+        Single<HttpResponse> result;
+
+        if (pollingSucceeded == null) {
+            updateDelayInMillisecondsFrom(httpPollResponse);
```

Comment:
in `updateFrom` sync method `updateDelayInMillisecondsFrom` is always called, but in async we do that only if `pollingSucceeded == null`, which one is correct?

Reply:
should we implement sync `updateFrom` using `updateFromAsync` (by blocking on it)?

Reply:
I didn't want to do that originally because of the comments that Jeffrey Richter made about synchronous code going through asynchronous code paths adds a lot of overhead. However, I think coding to avoid that right now adds a fair amount of maintenance (from the duplication). I'll Change updateFrom() to call updateFromAsync().